libs/libc/stdio/lib_open_memstream.c

```diff
@@ -0,0 +1,229 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <string.h>
+#include <assert.h>
+#include <errno.h>
+
+#include "libc.h"
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct memstream_cookie_s
+{
+  FAR char **buf;   /* Memory buffer */
+  char saved;       /* char at sizep before '\0' */
+  size_t *sizep;
+  off_t size;       /* Allocated buffer size */
+  off_t end;        /* Maximum position we have written to */
+  off_t pos;        /* Current position in the buffer */
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: memstream_write
+ ****************************************************************************/
+
+static ssize_t memstream_write(FAR void *c, FAR const char *buf,
+                               size_t size)
+{
+  FAR struct memstream_cookie_s *memstream_cookie =
+    (FAR struct memstream_cookie_s *)c;
+  FAR char *buf_grow;
+  size_t new_size;
+
+  if (memstream_cookie->pos + size + 1 > memstream_cookie->size)
+    {
+      /* We have to reallocate the buffer. */
+
+      new_size = memstream_cookie->pos + size + 1;
+      buf_grow = lib_realloc(*memstream_cookie->buf, new_size);
+      if (buf_grow == NULL)
+        {
+          return -ENOMEM;
+        }
+
+      memset(buf_grow + memstream_cookie->end, 0,
+            new_size - memstream_cookie->size);
+
+      *memstream_cookie->buf = buf_grow;
+      memstream_cookie->size = new_size;
+    }
+
+  memcpy(*memstream_cookie->buf + memstream_cookie->pos, buf, size);
+
+  /* POSIX: If a write moves the position to a value larger than
+   * the current length, the current length shall be set to this position.
+   * In this case a null character shall be appended to the current buffer.
+   */
+
+  memstream_cookie->pos += size;
+  if (memstream_cookie->pos > memstream_cookie->end)
+    {
+      memstream_cookie->end = memstream_cookie->pos;
+    }
+  else
+    {
+      memstream_cookie->saved = *(*memstream_cookie->buf +
```

Comment:
it is not untouched as we need to write `\0` at `memstream_cookie->pos`. And it may happen that we are for example in the middle of the buffer because of previous seeks so there might be useful data on `memstream_cookie->pos`.